pkg/authentication/auth0.go

```diff
@@ -0,0 +1,83 @@
+package authentication
+
+import (
+	"context"
+	"errors"
+	"github.com/golang-jwt/jwt/v5"
+	"strings"
+)
+
+var (
+	// ErrTokenNoProvided is returned when no token is provided.
+	ErrTokenNoProvided = errors.New("no token provided")
+	// ErrTokenInvalid is returned when a certain token is invalid.
+	ErrTokenInvalid = errors.New("token is not valid")
+	// ErrTokenMalformed is returned when the token doesn't match the expected token form.
+	// 	Example: JWT should have 3 parts, header, payload and signature.
+	ErrTokenMalformed = errors.New("token is malformed")
+
+	// ErrJWTNoHeader is returned when a certain JWT contains no header.
+	ErrJWTNoHeader = errors.New("jwt contains no header")
+
+	// ErrJWTNoPayload is returned when a certain JWT contains no payload.
+	ErrJWTNoPayload = errors.New("jwt contains no payload")
+
+	// ErrJWTNoSignature is returned when a certain JWt contains no signature.
+	ErrJWTNoSignature = errors.New("jwt contains no signature")
```

Comment:
We should create error definitions to encapsulate error types. The rationale is that code typically handles errors based on its type, rather than the specific error message within. I can see there being a benefit to having specific error messages for each returned error to help the user, but we can instead take the approach of having a small set of possible errors and wrapping them with specific error information.

In this case, I think it makes sense to only have `ErrTokenMissing` and `ErrTokenInvalid`, as the malformed or missing token segments errors are all invalid tokens. 

This change will help turn this

```go
if err == authentication.ErrTokenNoProvided {
  log.Error("No token was found")
  return
}
if err == authentication.TokenInvalid || err == authentication.ErrTokenMalformed || err == authentication.ErrJWTNoHeader || err == authentication.ErrJWTNoPayload || err == authentication.ErrJWTNoSignature {
  log.Error("The token is invalid:", err)
  return
}
```

into this

```go
if errors.Is(err, authentication.ErrTokenMissing) {
  log.Error("No token was found")
  return
}
if (err, authentication.ErrTokenInvalid) {
  log.Error("The token is invalid:", err)
  return
}
```

I'll admit this is a poor example because we don't really do anything different with the returned errors, which means that we will typically use catch-all `err != nil` error handling, but I hope this illustrates why having this many error definitions is cumbersome for the programmer when they want to catch errors, and not particularly useful. Having `VerifyJWT` wrap error types with specific information is just as useful to help the user.

Reply:
Sounds like a good suggestion. I implemented this change, care to give it another pass?

Reply:
Sounds like a good suggestion. I implemented this change, care to give it another pass?